deps: downgrade jruby, keep updated default-gem dependencies

yaauie opened the pull request:
## Release notes

[rn:skip] (the current PR for 8.10's release notes https://github.com/elastic/logstash/pull/15265/files contains no reference to the Jruby upgrade, so mentioning this downgrade which is still a net-upgrade from the one we shipped in LS 8.9 series doesn't make sense)

## What does this PR do?

Downgrades JRuby to 9.4.2.0 to avoid the silent global crash of the scheduler backing `Concurrent::TimerTask` that occurs when Jruby 9.4.3.0's invokedynamic promotes a method to run natively, incorrectly.

Upstream bug: https://github.com/jruby/jruby/issues/7904

Along with the downgrade of JRuby itself to 9.4.2.0, we cherry-pick the updates to default gems that were included in the latest JRuby 9.4.3.0 to ensure we don't back out relevant fixes to stdlib.

We also remove a pinned-dependency on `racc` that is no longer relevant, and pick up changes for the `Gemfile.*.lock.release` that _should have_ been there:
 - platform `universal-java-11` should not exist, as it is intentionally removed by the `setupJRuby` gradle task
 - `puma` changes in `logstash-core.gemspec` should be reflected in the release lockfile

Resolves: https://github.com/elastic/logstash/issues/15282

## Why is it important/What is the impact to the user?

By downgrading JRuby we can avoid a regression that causes silent failure of the globally-shared scheduler that powers a wide range of things from metrics captures for the API and the geoip database manager to the basic functionality of output plugins like `s3` and `http`.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

With a long-running pipeline and logging in debug mode, it is easy to see that the scheduled work keeps working when this patch is applied, or silently stops without this patch:

~~~
bin/logstash --log.level debug -e 'input { heartbeat {} }'
~~~

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseeds #123
-->
- 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

yaauie commented on Gemfile.jruby-3.1.lock.release:
```diff
@@ -8,28 +8,37 @@ PATH
   remote: logstash-core
   specs:
     logstash-core (8.10.0-java)
+      cgi (~> 0.3.6)
```
review note: default gems don't normally appear in the Gem's lockfile, but since we declare versioned dependencies in the `gemspec` they appear here in the resolved lock. When we upgrade to Jruby 9.4.x that includes them, they will be removed from the `gemspec`, which will cause them to disappear from this lock-file.

yaauie commented on Gemfile.jruby-3.1.lock.release:
```diff
@@ -8,28 +8,37 @@ PATH
+      net-protocol (~> 0.1.2)
       pry (~> 0.12)
-      puma (~> 6.3, >= 6.0.0)
-      racc (~> 1.5.2)
```
`racc` v 1.6.0 is a default gem in JRuby 9.4.2.0

yaauie commented on Gemfile.jruby-3.1.lock.release:
```diff
@@ -970,7 +993,7 @@ DEPENDENCIES
   polyglot
   rack-test
-  rake (~> 12)
+  rake (~> 13)
```
this should have been reflected after #15257

yaauie commented on Gemfile.jruby-3.1.lock.release:
```diff
@@ -852,7 +876,6 @@ GEM
 
 PLATFORMS
   java
-  universal-java-11
```
This should not have been included; unclear why it was.

yaauie reviewed (commented):
I have added review notes in-line

yaauie commented:
TODO: licenses for all new "default gems" listed in the lock file should be added to make the dependency report happy. I believe they all are the Ruby license.

kaisecheng commented:
I checked this branch and ran `./gradlew clean bootstrap assemble installDefaultGems` successfully but failed to `./gradlew test`

```
> Task :bootstrap
2023-08-29T14:48:02.989+01:00 [Execution worker Thread 4] WARN FilenoUtil : Native subprocess control requires open access to the JDK IO subsystem
Pass '--add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.base/java.io=ALL-UNNAMED' to enable.
Error Errno::ENOENT, retrying 1/10
No such file or directory - /var/folders/s3/bmfwqvs17pjb59crqdbgq_n40000gn/T/bin/jruby
Error Errno::ENOENT, retrying 2/10
No such file or directory - /var/folders/s3/bmfwqvs17pjb59crqdbgq_n40000gn/T/bin/jruby
...
Error Errno::ENOENT, retrying 10/10
No such file or directory - /var/folders/s3/bmfwqvs17pjb59crqdbgq_n40000gn/T/bin/jruby
Too many retries, aborting, caused by Errno::ENOENT

> Task :bootstrap FAILED

FAILURE: Build failed with an exception.

* Where:
Script '/Users/kcheng/GitHub/kaisecheng/logstash/rubyUtils.gradle' line: 174

* What went wrong:
Execution failed for task ':bootstrap'.
> (ENOENT) No such file or directory - /var/folders/s3/bmfwqvs17pjb59crqdbgq_n40000gn/T/bin/jruby
```

Are you able to run `./gradlew test` locally?

yaauie commented:
> No such file or directory - /var/folders/s3/bmfwqvs17pjb59crqdbgq_n40000gn/T/bin/jruby

I ran into this too, and am not 100% sure how I got past it.

Running gradle with the `--no-daemon` flag seemed to make it use the "right" jruby, but I _think_ it is only trying to use that jruby to `Kernel::exec` to create a clean bunder env, which it does when the `Gemfile*.lock*` had a `BUNDLED WITH` clause that didn't match the version of bundler activated which caused me to hit https://github.com/elastic/logstash/issues/15284. I _believe_ I resolved both by manually editing the lock file, but may have also just deleted the lock files and allowed them to regenerate cleanly with `./gradlew --no-daemon clean assemble installDefaultGems`.

 while gradle was re-using a daemon that had , and after a lot of chasing (using `--stacktrace` gradle arg) I found that invoking once with `--no-daemon` seems to have resolved the issue.

yaauie commented:
jenkins test this again please

yaauie commented on tools/dependencies-report/src/main/resources/notices/ffi-binary-libfixposix-NOTICE.txt:
```diff
@@ -0,0 +1,277 @@
+Eclipse Public License - v 2.0
```
this gem is [tri-licensed (`Ruby`, `EPL-2.0`, `LGPL-2.1-or-later`)](https://github.com/byteit101/subspawn/blob/master/subspawn-posix/subspawn-posix.gemspec#L40), and the Ruby license is preferable to EPL so I will swap it out.

yaauie commented:
Note: [Github webhooks are significantly degraded](https://www.githubstatus.com/incidents/wqv4yvlgtq62), and we are still waiting for builds to get kicked off. No major code-changes since last green build ([`58a9baf`..`yaauie:downgrade-jruby-8-10`](https://github.com/yaauie/logstash/compare/58a9baf..downgrade-jruby-8-10))

kaisecheng reviewed (approved):
LGTM. The scheduler works as expected and the `--no-daemon` allows tests to run. CI is green

mashhurs commented on logstash-core/logstash-core.gemspec:
```diff
@@ -86,7 +86,15 @@ Gem::Specification.new do |gem|
+  gem.add_runtime_dependency "ffi-binary-libfixposix", "~> 0.5.1.1"
+  gem.add_runtime_dependency "time", "~> 0.2.2"
+  gem.add_runtime_dependency "timeout", "~> 0.3.2"
+  gem.add_runtime_dependency "uri", "~> 0.12.1"
```
I have run several configs, includes s3-input, http-input, tcp-input, etc.. and I didn't face any issues.
However, I am a bit curious about these declarations since I don't know internal details. 
Can I know based on what assumptions/references we added?

It seems to me scheduler doesn't have direct/indirect `net-http` dependency and test by removing it gives also successful result. Additionally, `net-http` seems internally includes `uri` (based on [this explanation](https://github.com/ruby/net-http#usage)) which may not explicitly define?

yaauie replied:
Much of what was once the ruby standard library is now made available as separately-versioned gems, and ruby distributions like jruby now include "default gems" so that when a fix is made available in one component of the stdlib the user doesn't have to upgrade their whole ruby distribution — they can just install the individual gem and it will be used instead of its "default gems" counterpart.

We get these default gems "for free" by just using ruby (or jruby), so they aren't normally included in the explicit dependencies.

Jruby 9.4.3.0 includes a bunch of updates to these default gems over those shipped with the 9.4.2.0 that we are downgrading to. Declaring the updated versions here ensures that we have the most up-to-date versions including security fixes etc., and it also means that they become an explicit part of the dependency graph that is reflected in the generated lockfile. This is temporary, and can go away once the upstream issue is fixed.

robbavey reviewed (approved):
Tested - thread death failures observed in `main` do not occur here.

One question on whether the `jruby-openssl` restriction is temporary, and should be commented as such

yaauie commented on logstash-core/logstash-core.gemspec:
```diff
@@ -56,7 +56,7 @@ Gem::Specification.new do |gem|
   gem.add_runtime_dependency "sinatra", '~> 2'
   gem.add_runtime_dependency 'puma', '~> 6.3', '>= 6.3.1'
-  gem.add_runtime_dependency "jruby-openssl", "~> 0.11"
+  gem.add_runtime_dependency "jruby-openssl", "~> 0.14.1"
```
No. The low-pin used to be an "upgrade" that got stuck and is now behind what is shipped with jruby (see git blame)

mashhurs reviewed (commented):
LGTM!
